### ISSUE TITLE
new-IR lacks a Basic-DCE (dead code elimination) pass

### ISSUE THREAD

**Issue by vchuravy** [JuliaLang/julia#28212]
```
@inline function f(A, I)
    @boundscheck checkbounds(A, I)
    @inbounds A[I]
end
g(A, I) = @inbounds f(A, I)
```

On `0.7-beta2` `@code_typed g(ones(1), 1)` will contain  a `checkbounds` call.

cc: @mbauman

**Comment by mbauman** [comment]
For others here: note that the `@code_llvm g(ones(1), 1)` will show that it has removed the bounds check, but that's only because `Array` has special bounds-check elision.  Do it with any custom array type and the bounds-checks are not removed within LLVM, either.

**Comment by vtjnash** [comment]
> 1 1 ─       goto 6 if not false

Looks like the basic-dce pass hasn't been reimplemented yet in the new optimizer. That'll eventually happen at the LLVM level, although it might spoil our early alloc-elim passes at the Julia level.

**Comment by Keno** [comment]
dup of #27547 for the Julia IR parts of this. I have most of a branch that fixes that, but I basically only intend to get the #28211 parts of that merged for 0.7 since I want to avoid introducing additional instability.

**Comment by mbauman** [comment]
We had been trying to track down another perf bug and this caught us off-guard… but it turned out to be something else entirely.  And it looks like I had been doing something funny in my checks with custom arrays — those get eliminated by LLVM, too, it seems.

So I'm not sure this actually is harmful beyond being surprising in the Julia IR.
